File: src/main/webapp/resources/core/js/vector2d.js
"use strict";

var game = game || {};

game.vector2d = function(spec){
    var that = {};

    var x = spec.x;
    var y = spec.y;

    that.getX = function() {
        return x;
    };

    that.getY = function() {
        return y;
    };

    that.add = function(that) {
        x += that.getX();
        y += that.getY();
    };

    that.addX = function(newX) {
        x += newX;
    };

    that.addY = function(newY) {
        y += newY;
    };

    return that;
}

File: src/main/java/service/Vector2d.java
package service;

public class Vector2d {
    private double x;
    private double y;

    public static Vector2d zero(){
        return new Vector2d(0, 0);
    }

    public Vector2d(double x, double y) {
        this.x = x;
        this.y = y;
    }

    public double x() {
        return x;
    }

    public double y() {
        return y;
    }

    public void scale(double scale) {
        x *= scale;
        y *= scale;
    }

    public void add(Vector2d that) {
        this.x += that.x;
        this.y += that.y;
    }
}
